lib/src/validator/leak_detection.dart

```diff
@@ -0,0 +1,799 @@
+
+  /// Patterns for file-paths where in matches for [_pattern] should be ignored.
+  final List<Pattern> _ignoredFiles;
+
+  /// Test vectors that will have a match in [findPossibleLeaks].
+  @visibleForTesting
+  final List<String> testsWithLeaks;
+
+  /// Test vectors that will not have a match in [findPossibleLeaks].
+  @visibleForTesting
+  final List<String> testsWithNoLeaks;
+
+  LeakPattern._({
+    @required this.kind,
+    @required String pattern,
+    Iterable<Pattern> allowed = const <Pattern>[],
+    Map<int, double> entropyThresholds = const <int, double>{},
+    Iterable<Pattern> allowedPreceeding = const <Pattern>[],
+    Iterable<Pattern> ignoredFiles = const <Pattern>[],
+    Iterable<String> testsWithLeaks = const <String>[],
+    Iterable<String> testsWithNoLeaks = const <String>[],
+  })  : _pattern = RegExp(pattern),
+        _allowed = List.unmodifiable(allowed),
+        _entropyThresholds = Map.unmodifiable(entropyThresholds),
+        _allowedPreceeding = List.unmodifiable(allowedPreceeding),
+        _ignoredFiles = List.unmodifiable(ignoredFiles),
+        testsWithLeaks = List.unmodifiable(testsWithLeaks),
+        testsWithNoLeaks = List.unmodifiable(testsWithNoLeaks);
+
+  /// Find possible leaks using this [LeakPattern].
+  ///
+  /// A possible [LeakMatch] is found when:
+  ///  * [_pattern] is matched,
+  ///  * no pattern in [_allowed] is matched,
+  ///  * 10 preceeding lines does not match any pattern in [_allowedPreceeding],
+  ///  * Captured group have a entropy higher than [_entropyThresholds] requires
+  ///    for the given _group identifier_, and,
+  ///  * [file] does not match any pattern in [_ignoredFiles].
+  Iterable<LeakMatch> findPossibleLeaks(String file, String content) sync* {
+    // Skip if we have a filename match
+    if (_ignoredFiles.any((p) => p.allMatches(file).isNotEmpty)) {
+      return;
+    }
+    final source = SourceFile.fromString(content, url: file);
+    for (final m in _pattern.allMatches(content)) {
+      if (_allowed.any((s) => m.group(0).contains(s))) {
+        continue;
+      }
+      if (_entropyThresholds.entries
+          .any((entry) => _entropy(m.group(entry.key)) < entry.value)) {
+        continue;
+      }
+      // Find text from 10 lines preceeding the match
+      var offset = m.start;
+      for (var i = 0; i < 10 && offset > 0; i++) {
+        offset = content.lastIndexOf('\n', offset - 1);
+        if (offset <= 0) {
+          offset = 0;
+        }
+      }
+      final preceedingLines = content.substring(offset, m.start);
+      if (_allowedPreceeding.any(preceedingLines.contains)) {
+        continue;
+      }
+
+      yield LeakMatch(
+        this,
+        source.span(m.start, m.start + m.group(0).length),
+      );
+    }
+  }
+}
+
+/// Compute Shannon entropy [1] of [s].
+///
+/// [1]: https://en.wikipedia.org/w/index.php?title=Entropy_(information_theory)&oldid=1033726547
+double _entropy(String s) {
```

Comment:
I believe there is no such thing as the entropy of a string, only of a distribution...
https://stackoverflow.com/a/2998054
Not sure exactly what we are trying to do here...